File: backend/app/schemas/department.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class DepartmentProfileResponse(BaseModel):
    id: UUID
    user_id: UUID
    tpo_profile_id: UUID
    officer_name: str
    officer_email: str
    officer_contact: str
    department_name: str

    model_config = ConfigDict(from_attributes=True)


class DepartmentStudentResponse(BaseModel):
    user_id: UUID
    profile_id: UUID
    full_name: str
    college_email: str
    personal_email: str
    phone_number: str
    date_of_birth: str
    gender: str
    exam_roll_number: str
    degree: str
    batch_year: int
    graduation_year: int


class DepartmentTestResponse(BaseModel):
    id: UUID
    title: str
    description: str | None
    mode: str
    status: str
    duration_minutes: int | None
    created_at: datetime
    released_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class DepartmentDashboardResponse(BaseModel):
    total_students: int
    total_tests: int
    active_tests: int
    completed_tests: int
    pending_result_release: int


from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class DepartmentProfileResponse(BaseModel):
    id: UUID
    user_id: UUID
    department_name: str

    class Config:
        from_attributes = True


class DepartmentStudentResponse(BaseModel):
    user_id: UUID
    profile_id: UUID
    full_name: str
    college_email: str
    personal_email: str | None = None
    phone_number: str | None = None
    date_of_birth: str | None = None
    gender: str | None = None
    exam_roll_number: str | None = None
    degree: str | None = None
    batch_year: int | None = None
    graduation_year: int | None = None


class DepartmentStudentTestRecord(BaseModel):
    assignment_id: UUID
    test_id: UUID
    test_title: str
    test_mode: str
    status: str
    started_at: datetime | None = None
    submitted_at: datetime | None = None
    score: float | None = None
    percentage: float | None = None


class DepartmentStudentDetailResponse(BaseModel):
    student: DepartmentStudentResponse
    test_history: list[DepartmentStudentTestRecord]
    performance: dict


class DepartmentTestResponse(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    mode: str
    status: str
    duration_minutes: int | None = None
    created_at: datetime
    released_at: datetime | None = None


class DepartmentTestDetailResponse(BaseModel):
    id: UUID
    title: str
    description: str | None = None
    mode: str
    status: str
    duration_minutes: int | None = None
    created_at: datetime
    released_at: datetime | None = None

    question_count: int
    assigned_count: int
    in_progress_count: int
    submitted_count: int
    missed_count: int


class DashboardTestSummary(BaseModel):
    id: UUID
    title: str
    mode: str
    status: str
    created_at: datetime
    released_at: datetime | None = None
    assigned_count: int
    submitted_count: int
    missed_count: int


class DepartmentDashboardResponse(BaseModel):
    total_students: int
    total_tests: int

    draft_tests: int
    released_tests: int
    closed_tests: int

    total_assignments: int
    submitted_assignments: int
    in_progress_assignments: int
    missed_assignments: int

    average_percentage: float | None = None

    recent_tests: list[DashboardTestSummary]


class DepartmentTestMonitorStudent(BaseModel):
    assignment_id: UUID
    student_profile_id: UUID
    student_name: str
    status: str
    started_at: datetime | None = None
    submitted_at: datetime | None = None


class DepartmentTestMonitorResponse(BaseModel):
    test_id: UUID
    test_title: str

    assigned_count: int
    in_progress_count: int
    submitted_count: int
    missed_count: int

    students: list[DepartmentTestMonitorStudent]


class DepartmentTestResultStudent(BaseModel):
    assignment_id: UUID
    student_profile_id: UUID
    student_name: str

    total_marks: float | None = None
    obtained_marks: float | None = None
    percentage: float | None = None

    llm_analysis_completed: bool
    released: bool
    released_at: datetime | None = None


class DepartmentTestResultsResponse(BaseModel):
    test_id: UUID
    test_title: str
    results: list[DepartmentTestResultStudent]

File: frontend/src/api/department.ts
import { apiRequest } from "./client";

/* =========================================================
   Shared Types
========================================================= */

export type TestMode =
  | "AUTOMATIC"
  | "MANUAL"
  | "ADAPTIVE";

export type TestStatus =
  | "DRAFT"
  | "RELEASED"
  | "CLOSED";

export type AssignmentStatus =
  | "ASSIGNED"
  | "IN_PROGRESS"
  | "SUBMITTED"
  | "MISSED";


/* =========================================================
   Department Profile
========================================================= */

export interface DepartmentProfile {
  id: string;
  user_id: string;
  tpo_profile_id: string;

  officer_name: string;
  officer_email: string;
  officer_contact: string;

  department_name: string;
}


/* =========================================================
   Dashboard
========================================================= */

export interface DashboardTestSummary {
  id: string;
  title: string;

  mode: TestMode;
  status: TestStatus;

  created_at: string;
  released_at?: string | null;

  assigned_count: number;
  submitted_count: number;
  missed_count: number;
}


export interface DepartmentDashboardResponse {
  total_students: number;
  total_tests: number;

  draft_tests: number;
  released_tests: number;
  closed_tests: number;

  total_assignments: number;
  submitted_assignments: number;
  in_progress_assignments: number;
  missed_assignments: number;

  average_percentage: number | null;

  recent_tests: DashboardTestSummary[];
}


/* =========================================================
   Students
========================================================= */

export interface DepartmentStudent {
  user_id: string;
  profile_id: string;

  full_name: string;

  college_email: string;
  personal_email?: string | null;

  phone_number?: string | null;

  date_of_birth?: string | null;
  gender?: string | null;

  exam_roll_number?: string | null;

  degree?: string | null;

  batch_year?: number | null;
  graduation_year?: number | null;
}


/* =========================================================
   Student Test History
========================================================= */

export interface DepartmentStudentTestRecord {
  assignment_id: string;
  test_id: string;

  test_title: string;

  test_mode: TestMode;

  status: AssignmentStatus;

  started_at?: string | null;
  submitted_at?: string | null;

  score?: number | null;
  percentage?: number | null;
}


/* =========================================================
   Student Performance
========================================================= */

export interface DepartmentStudentPerformance {
  total_tests: number;

  completed_tests: number;
  in_progress_tests: number;
  missed_tests: number;

  average_percentage?: number | null;
}


/* =========================================================
   Student Detail
========================================================= */

export interface DepartmentStudentDetail {
  student: DepartmentStudent;

  test_history: DepartmentStudentTestRecord[];

  performance: DepartmentStudentPerformance;
}


/* =========================================================
   Tests
========================================================= */

export interface DepartmentTest {
  id: string;

  title: string;
  description?: string | null;

  mode: TestMode;
  status: TestStatus;

  duration_minutes?: number | null;

  created_at: string;
  released_at?: string | null;
}


/* =========================================================
   Test Detail
========================================================= */

export interface DepartmentTestDetail {
  id: string;

  title: string;
  description?: string | null;

  mode: TestMode;
  status: TestStatus;

  duration_minutes?: number | null;

  created_at: string;
  released_at?: string | null;

  question_count: number;

  assigned_count: number;
  in_progress_count: number;
  submitted_count: number;
  missed_count: number;
}


/* =========================================================
   Test Monitor
========================================================= */

export interface DepartmentTestMonitorStudent {
  assignment_id: string;
  student_profile_id: string;

  student_name: string;

  status: AssignmentStatus;

  started_at?: string | null;
  submitted_at?: string | null;
}


export interface DepartmentTestMonitorResponse {
  test_id: string;
  test_title: string;

  assigned_count: number;
  in_progress_count: number;
  submitted_count: number;
  missed_count: number;

  students: DepartmentTestMonitorStudent[];
}


/* =========================================================
   Test Results
   Kept here for later.
========================================================= */

export interface DepartmentTestResultStudent {
  assignment_id: string;
  student_profile_id: string;

  student_name: string;

  total_marks?: number | null;
  obtained_marks?: number | null;
  percentage?: number | null;

  llm_analysis_completed: boolean;

  released: boolean;
  released_at?: string | null;
}


export interface DepartmentTestResultsResponse {
  test_id: string;
  test_title: string;

  results: DepartmentTestResultStudent[];
}


/* =========================================================
   API FUNCTIONS
========================================================= */


/* ---------- Profile ---------- */

export function getDepartmentProfile() {
  return apiRequest<DepartmentProfile>(
    "/api/department/me",
  );
}


/* ---------- Dashboard ---------- */

export function getDepartmentDashboard() {
  return apiRequest<DepartmentDashboardResponse>(
    "/api/department/dashboard",
  );
}


/* ---------- Students ---------- */

export function getDepartmentStudents() {
  return apiRequest<DepartmentStudent[]>(
    "/api/department/students",
  );
}


export function getDepartmentStudentDetail(
  studentId: string,
) {
  return apiRequest<DepartmentStudentDetail>(
    `/api/department/students/${studentId}`,
  );
}


/* ---------- Tests ---------- */

export function getDepartmentTests() {
  return apiRequest<DepartmentTest[]>(
    "/api/department/tests",
  );
}


export function getDepartmentTestDetail(
  testId: string,
) {
  return apiRequest<DepartmentTestDetail>(
    `/api/department/tests/${testId}`,
  );
}


/* ---------- Test Monitor ---------- */

export function getDepartmentTestMonitor(
  testId: string,
) {
  return apiRequest<DepartmentTestMonitorResponse>(
    `/api/department/tests/${testId}/monitor`,
  );
}


/* ---------- Test Results ---------- */

export function getDepartmentTestResults(
  testId: string,
) {
  return apiRequest<DepartmentTestResultsResponse>(
    `/api/department/tests/${testId}/results`,
  );
}